### PR TITLE
Convert frontend-block.js to frontend-block.tsx and replace propTypes with TypeScript definitions 

### LINKED ISSUE #9525: Convert `frontend-block.js` to `frontend-block.tsx` and replace `propTypes` with TypeScript definitions
## Problem

[`assets/js/blocks/reviews/frontend-block.js`](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/blocks/reviews/frontend-block.js) is a JS file and uses `propTypes` definitions.

## Expected outcome

1. Convert the file from `frontend-block.js` to `frontend-block.tsx`.
2. Replace `propTypes` with TypeScript definitions.

### PR DESCRIPTION
**Pull request by samueljseay**
## What

Fixes #9525 and while in the area cleans up some use of generics, makes attributes types more specific and adds some supporting tests for the component being refactored.

## Testing Instructions

Apart from the tests passing you could also smoke test the reviews block in store front-end and see that it still works as expected (I've tested this).

* [x] Do not include in the Testing Notes 
* [x] Should be tested by the development team exclusively 

## WooCommerce Visibility

Required:

* [ ] WooCommerce Core
* [x] Feature plugin
* [ ] Experimental
* [ ] N/A

## Checklist

Required:
* [x] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [ ] This PR is assigned to a milestone.

Conditional:
* [ ] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

## Changelog

> Skipping

### CODE REVIEW

File: assets/js/blocks/reviews/frontend-block.tsx [hunk 1]
```diff
@@ -2,23 +2,43 @@
...
 
 /**
  * Block rendered in the frontend.
  *
- * @param {Object}                                             props                 Incoming props for the component.
- * @param {Object}                                             props.attributes      Incoming block attributes.
- * @param {function(any):any}                                  props.onAppendReviews Function called when appending review.
- * @param {function(any):any}                                  props.onChangeOrderby
- * @param {Array}                                              props.reviews
+ * @param {FrontendBlockProps}                                 props                 Incoming props for the component.
```
Comment (samueljseay) [inline assets/js/blocks/reviews/frontend-block.tsx]: I cleaned up the JS doc but looking around the repo, I'm not sure if anyone is using JS Doc right now?

File: assets/js/base/components/reviews/review-list/index.tsx [hunk 2]
```diff
@@ -11,17 +11,30 @@ import ReviewListItem from '../review-list-item';
 import type { Review } from '../types';
 import './style.scss';
 
+export type ReviewListAttributes = BlockAttributes & {
+	imageType: string;
```
Comment (samueljseay) [inline assets/js/base/components/reviews/review-list/index.tsx]: One thing I'm not sure about is how we should type the block attributes,

With blocks I think attributes should all be `?` technically, but if we pass a lot of fallback defaults for attributes maybe these will always be present

File: assets/js/blocks/reviews/frontend-block.tsx [hunk 3]
```diff
@@ -2,23 +2,43 @@
...
+import type { ChangeEventHandler, MouseEventHandler } from 'react';
+
+/**
+ * Internal dependencies
+ */
+import { ReviewListAttributes } from '../../base/components/reviews/review-list';
+import { Review } from '../../base/components/reviews/types';
+
+type FrontendBlockAttributes = ReviewListAttributes & {
+	showOrderby?: 'false' | 'true';
```
Comment (samueljseay) [inline assets/js/blocks/reviews/frontend-block.tsx]: These also I couldn't decide should they be `?` or not

File: assets/js/blocks/reviews/frontend-block.tsx (hunk 1 above)
Comment (dinhtungdu) [reply assets/js/blocks/reviews/frontend-block.tsx]: Personally, I also don't use JS Doc for ts/tsx files.

File: assets/js/base/components/reviews/review-list/index.tsx (hunk 2 above)
Comment (dinhtungdu) [reply assets/js/base/components/reviews/review-list/index.tsx]: I think we should type those attributes according to how we register them in `assets/js/blocks/reviews/attributes.js`.

By the way, I also think we should export the attribute type from the block directory then import it here, it's easier to update the type in the future.

File: assets/js/blocks/reviews/frontend-block.tsx (hunk 3 above)
Comment (dinhtungdu) [reply assets/js/blocks/reviews/frontend-block.tsx]: I think the `showOrderby` and `showLoadMore` attributes aren't prepared properly, leading them to hold string values (`'true'` and `'false'`) when the block is rendered on the frontend. Those attributes are typed as boolean according to `assets/js/blocks/reviews/attributes.js`.

IMO, we should update `assets/js/blocks/reviews/frontend.js` to process the value of those attributes to match their registered types.

**Review by dinhtungdu — COMMENTED** [review]
Thank you so much for working on this @samueljseay! And thanks for adding the tests. The PR looks good, review blocks are working as they are before. I left some comments on your PR, can you please take a look?

File: assets/js/blocks/reviews/frontend-block.tsx (hunk 3 above)
Comment (dinhtungdu) [reply assets/js/blocks/reviews/frontend-block.tsx]: You can take the Mini Cart block as [an example](https://github.com/woocommerce/woocommerce-blocks/blob/6248528985fde1900a1631cb7c74fe00b4eb6f2e/assets/js/blocks/mini-cart/component-frontend.tsx#L47).

File: assets/js/blocks/reviews/frontend-block.tsx (hunk 1 above)
Comment (samueljseay) [reply assets/js/blocks/reviews/frontend-block.tsx]: Thanks, removed in [6248528](https://github.com/woocommerce/woocommerce-blocks/pull/10677/commits/6248528985fde1900a1631cb7c74fe00b4eb6f2e)

File: assets/js/blocks/reviews/frontend.js [hunk 4]
```diff
@@ -15,8 +15,13 @@ const selector = `
 `;
 
 const getProps = ( el ) => {
+	const showOrderby = el.dataset.showOrderby === 'true';
```
Comment (samueljseay) [inline assets/js/blocks/reviews/frontend.js:18]: handles undefined, 'false' and any string value that isn't 'true' | 'false'

**Comment by samueljseay** [comment]
Thanks for the review @dinhtungdu when you get a chance could you take another look,

I've moved the attribute type into attributes so it can be easily kept up to date, I've cooerced those 2 attributes to boolean and treated them accordingly in the component. The tests cover the change to boolean of the attributes too.

File: assets/js/base/components/reviews/review-list-item/index.tsx [hunk 5]
```diff
@@ -4,13 +4,13 @@
...
 import classNames from 'classnames';
 import ReadMore from '@woocommerce/base-components/read-more';
-import type { BlockAttributes } from '@wordpress/blocks';
 
 /**
  * Internal dependencies
  */
 import './style.scss';
 import type { Review } from '../types';
+import { ReviewBlockAttributes } from '../../../../blocks/reviews/attributes';
```
Comment (dinhtungdu) [inline assets/js/base/components/reviews/review-list-item/index.tsx]: For long relative imports, we can use alias instead. You can see the full list of aliases [here](https://github.com/woocommerce/woocommerce-blocks/blob/1d0af208720498af118e76f9e0e3c641f708b1e0/bin/webpack-helpers.js#L35-L99).

File: assets/js/base/components/reviews/review-list/index.tsx [hunk 6]
```diff
@@ -2,26 +2,28 @@
...
-import type { BlockAttributes } from '@wordpress/blocks';
 
 /**
  * Internal dependencies
  */
 import ReviewListItem from '../review-list-item';
 import type { Review } from '../types';
 import './style.scss';
-
+import { ReviewBlockAttributes } from '../../../../blocks/reviews/attributes';
```
Comment (dinhtungdu) [inline assets/js/base/components/reviews/review-list/index.tsx]: Long relative import here.

File: assets/js/blocks/reviews/attributes.ts [hunk 7]
```diff
@@ -1,3 +1,13 @@
+/**
+ * External dependencies
+ */
+import { BlockAttributes } from '@wordpress/blocks';
+
+/**
+ * Internal dependencies
+ */
+import { Review } from '../../base/components/reviews/types';
```
Comment (dinhtungdu) [inline assets/js/blocks/reviews/attributes.ts]: Long relative import here.

File: assets/js/blocks/reviews/frontend-block.tsx [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+	ReviewList,
+	ReviewSortSelect,
+} from '@woocommerce/base-components/reviews';
+import withReviews from '@woocommerce/base-hocs/with-reviews';
+import type { ChangeEventHandler, MouseEventHandler } from 'react';
+
+/**
+ * Internal dependencies
+ */
+import { Review } from '../../base/components/reviews/types';
```
Comment (dinhtungdu) [inline assets/js/blocks/reviews/frontend-block.tsx]: Long relative import here.

**Review by dinhtungdu — COMMENTED** [review]
@samueljseay thanks for the update, just some minor comments about relative import, otherwise, this is LGTM!

File: assets/js/blocks/reviews/frontend-block.tsx (hunk 8 above)
Comment (samueljseay) [reply assets/js/blocks/reviews/frontend-block.tsx]: Resolved in https://github.com/woocommerce/woocommerce-blocks/pull/10677/commits/d836711e60966b62a986c1d5a0e2aa82386cd181

File: assets/js/blocks/reviews/attributes.ts (hunk 7 above)
Comment (samueljseay) [reply assets/js/blocks/reviews/attributes.ts]: Resolved in https://github.com/woocommerce/woocommerce-blocks/pull/10677/commits/f20885e4c2180c190b960ad04a02e943fbccd112

File: assets/js/base/components/reviews/review-list/index.tsx (hunk 6 above)
Comment (samueljseay) [reply assets/js/base/components/reviews/review-list/index.tsx]: Resolved in https://github.com/woocommerce/woocommerce-blocks/pull/10677/commits/f20885e4c2180c190b960ad04a02e943fbccd112

File: assets/js/base/components/reviews/review-list-item/index.tsx (hunk 5 above)
Comment (samueljseay) [reply assets/js/base/components/reviews/review-list-item/index.tsx]: Resolved in https://github.com/woocommerce/woocommerce-blocks/commit/f20885e4c2180c190b960ad04a02e943fbccd112

**Comment by samueljseay** [comment]
Thanks @dinhtungdu , addressed your comments about the alias 😄

**Review by dinhtungdu — APPROVED** [review]
Thank for being patient with me! Great work! 🚀

**Pull request merged**